cmake/compiler/gcc/target_x86.cmake

```diff
@@ -8,6 +8,11 @@ else()
   string(PREPEND CMAKE_ASM_FLAGS "-m32 ")
   string(PREPEND CMAKE_C_FLAGS   "-m32 ")
   string(PREPEND CMAKE_CXX_FLAGS "-m32 ")
+
+  if(CONFIG_FP_USE_SOFT_FLOAT)
+    list(APPEND TOOLCHAIN_C_FLAGS  -msoft-float)
+    list(APPEND TOOLCHAIN_LD_FLAGS -msoft-float)
```

Comment:
This is needed for GCC to find the correct library (when calling GCC with `--print-libgcc-file-name`). Or else it would use the default library which uses hardware FPU.

Reply:
> This is needed for GCC to find the correct library (when calling GCC with `--print-libgcc-file-name`). Or else it would use the default library which uses hardware FPU.

Interesting, this only seems to be needed for 32 bit when using Zephyr SDK.

Distro gcc:
32 bit:
```
$ gcc-8 --print-libgcc-file-name -Wa,--divide -m32
/usr/lib/gcc/x86_64-linux-gnu/8/32/libgcc.a
$ gcc-8 --print-libgcc-file-name -Wa,--divide -m32 -msoft-float
/usr/lib/gcc/x86_64-linux-gnu/8/32/libgcc.a
```

64 bit:
```
$ gcc-8 --print-libgcc-file-name -Wa,--divide -m64 
/usr/lib/gcc/x86_64-linux-gnu/8/libgcc.a
$ gcc-8 --print-libgcc-file-name -Wa,--divide -m64 -msoft-float 
/usr/lib/gcc/x86_64-linux-gnu/8/libgcc.a
```

Zephyr SDK gcc, x86 (which this code affects):
64 bit:
```
$ /opt/zephyr-sdk-0.12.0-beta/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-gcc --print-libgcc-file-name -m64 
/opt/zephyr-sdk-0.12.0-beta/x86_64-zephyr-elf/bin/../lib/gcc/x86_64-zephyr-elf/10.2.0/libgcc.a
$ /opt/zephyr-sdk-0.12.0-beta/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-gcc --print-libgcc-file-name -m64 -msoft-float 
/opt/zephyr-sdk-0.12.0-beta/x86_64-zephyr-elf/bin/../lib/gcc/x86_64-zephyr-elf/10.2.0/libgcc.a
```
32 bit:
```
$ /opt/zephyr-sdk-0.12.0-beta/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-gcc --print-libgcc-file-name -m32 
/opt/zephyr-sdk-0.12.0-beta/x86_64-zephyr-elf/bin/../lib/gcc/x86_64-zephyr-elf/10.2.0/32/libgcc.a
$ /opt/zephyr-sdk-0.12.0-beta/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-gcc --print-libgcc-file-name -m32 -msoft-float 
/opt/zephyr-sdk-0.12.0-beta/x86_64-zephyr-elf/bin/../lib/gcc/x86_64-zephyr-elf/10.2.0/32/soft-float/libgcc.a
```

But things are a little different for arm:
Zephyr SDK gcc, arm:
```
$ /opt/zephyr-sdk-0.12.0-beta/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc --print-libgcc-file-name -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfpu=fpv4-sp-d16 -mfloat-abi=softfp
/opt/zephyr-sdk-0.12.0-beta/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.2.0/thumb/v7e-m+fp/softfp/libgcc.a
$ /opt/zephyr-sdk-0.12.0-beta/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc --print-libgcc-file-name -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfpu=fpv4-sp-d16 -mfloat-abi=softfp -msoft-float
/opt/zephyr-sdk-0.12.0-beta/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.2.0/libgcc.a
$ /opt/zephyr-sdk-0.12.0-beta/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc --print-libgcc-file-name -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfpu=fpv4-sp-d16 -mfloat-abi=soft 
/opt/zephyr-sdk-0.12.0-beta/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.2.0/thumb/v7e-m/nofp/libgcc.a
$ /opt/zephyr-sdk-0.12.0-beta/arm-zephyr-eabi/bin/arm-zephyr-eabi-gcc --print-libgcc-file-name -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfpu=fpv4-sp-d16 -mfloat-abi=soft -msoft-float
/opt/zephyr-sdk-0.12.0-beta/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.2.0/thumb/v7e-m/nofp/libgcc.a
```
which shows that we should be careful on combining `-mfloat-abi=softfp` with `-msoft-float`.

GNU ARM embedded, gcc:
```
$ arm-none-eabi-gcc --print-libgcc-file-name -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfpu=fpv4-sp-d16 -mfloat-abi=soft 
/opt/gcc-arm-none-eabi-7-2018-q2-update/bin/../lib/gcc/arm-none-eabi/7.3.1/libgcc.a
$ arm-none-eabi-gcc --print-libgcc-file-name -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfpu=fpv4-sp-d16 -mfloat-abi=soft -msoft-float
/opt/gcc-arm-none-eabi-7-2018-q2-update/bin/../lib/gcc/arm-none-eabi/7.3.1/libgcc.a
$ arm-none-eabi-gcc --print-libgcc-file-name -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfpu=fpv4-sp-d16 -mfloat-abi=softfp
/opt/gcc-arm-none-eabi-7-2018-q2-update/bin/../lib/gcc/arm-none-eabi/7.3.1/thumb/v7e-m/fpv4-sp/softfp/libgcc.a
$ arm-none-eabi-gcc --print-libgcc-file-name -mcpu=cortex-m4 -mthumb -mabi=aapcs -mfpu=fpv4-sp-d16 -mfloat-abi=softfp -msoft-float
/opt/gcc-arm-none-eabi-7-2018-q2-update/bin/../lib/gcc/arm-none-eabi/7.3.1/thumb/v7e-m/fpv4-sp/softfp/libgcc.a
```

The `TOOLCHAIN_C_FLAGS` are automatically added to the compile options here:
https://github.com/zephyrproject-rtos/zephyr/blob/fb99ef639a3237e83ee02664ac9211401c946099/CMakeLists.txt#L215-L217

Based on those observations, then we should either restrict the `CONFIG_FP_USE_SOFT_FLOAT` to the x86 platform only, as it is not enough that the toolchain supports the flag.
We also need to ensure that the corresponding `target_<platform>.cmake` is updated accordingly, or move this code to https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/compiler/gcc/target.cmake to ensure consistent behavior on all platforms when using `CONFIG_FP_USE_SOFT_FLOAT=y`.

In this case, gcc doesn't need to have `-msoft-float` in https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/compiler/gcc/compiler_flags.cmake, as it will automatically be applied to compile flags.
Other toolchains might still want to use the template, as they probably won't need to obtain lib path in similar way as we do for gcc.

Reply:
Linux distributions usually don't ship with soft-float libraries so that is expected. x86-64 has hardware FP support anyway so this is not needed. Though the output from ARM toolchain in SDK is not expected. So I have limited all this to just x86 32-bit to avoid unintended side effects on other architectures. We can always add support for other arch later if so desired.

Reply:
> Linux distributions usually don't ship with soft-float libraries so that is expected. 

ahh, so similar problem as with the `-m32` where you need to install the 32 bit libs in addition to the normal 64 bit libs.
It will still print the (wrong) gcc lib it will be using without errors/warnings, but fail at later build stage.
I assume the same will then happen in this case, a failure at link time.

> Though the output from ARM toolchain in SDK is not expected. 

Exactly that one surprised me most.

> We can always add support for other arch later if so desired.

yep, let's keep it simple for now.